torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp

```diff
@@ -189,7 +189,7 @@ class DoubleBufferLoopCloner : public kir::IrVisitor {
     // Main: 0 to (extent-1)
     // Epilogue: (extent-1) to extent
 
-    auto index = IrBuilder::create<Int>(c10::nullopt);
+    auto index = double_buffer_loop_->index();
```

Comment:
I'm not sure if this is fine. This clones a loop and the cloned loop uses the same Val for the loop index, meaning the Val is defined and used in multiple different scopes. I don't know anything that might break due to this, but this doesn't seem quite right.

Reply:
Mechanically in this case specifically, we have a few for loops that are attached to the same iterdomain. As far as our current definition of iterdomain goes, if we have scope conflict with these for loops, to the same iterdomain, it should very likely have been something not correct happening. 

But long term I do think it makes sense to allocate different index variables for these for loops but should pre-allocated from idgraph not created offline by the forloop.

Reply:
The fact that `IterDomain` and `kir::ForLoop` are not 1-to-1 mapping is, I think, fine. It doesn't break any mental model as far as I see. However, I think we should try to make the scoping of Vals as valid as possible. That would avoid future unexpected behavior and surprise. We don't do much analysis once Fusion is lowered to Kernel, so not a big deal today, but would it be too much to maintain mappings from IterDomain to multiple index Vals?

Reply:
It will not be too much effort and actually I do think it is important to formalize the one-to-many iterdomain to loop mapping in some place like idgraph. The ability to keep track of pipelining/skewing/peeling/rotation holistically on top of tiling is big deal as these patterns are so commonly seen. 

That said, the missing piece here is whether we want to formalize such thing as operations on iterdomains, like split and merge and still keep the iterdomain-to-loop a 1-to-1 relationship, or we are actually ok with scanning a single iterdomain with multiple loops, which is what this boils down to. It's an open question at the moment and my feeling is that operation on iterdomain approach fits better in our system.  

Meanwhile, It's just not an effort that I think would fit in the same PR : ) .

Reply:
I agree it's an open question, but currently and also in this PR, we are taking the second approach, and I think that's reasonable for now. 

I guess my concern is things like pipelining and skewing can be something important in the future, but it seems they are not really a priority for the near future, and so I'm assuming we would end up having this technical debt of inconsistent scoping for some time, i.e., until we decide to work on such advanced loop transformations. If we could avoid adding another technical debt right now, that seems more appropriate to me.

Reply:
Sure. I assume a short-term commitment to the one iterdomain to many for loop approach would be better than an undecided state. We could just revisit when we want to steer to the other way,

In this case, If I still want to holistically manage all the integers, I'd need to be able to differentiate the prolog, main and epilogs in a later phase, caching for loop pointers isn't reliable, so maybe just add another attribute to kir::ForLoop?

Reply:
That sounds reasonable. I don't think we want to have too many attributes in `kir::ForLoop`, but let's see how it goes, and if it's really becoming unmanageable, we should revisit.

Reply:
updated. they should now be different. resolving for now. 
Further cleanup possible given that the for loop attribute is now explicit. Will do follow ups.

Reply:
The update looks great. Thanks for cleaning it up!